In [48]:
# from google.colab import drive
# drive.mount('/content/drive')

### 0. Загрузка данных

In [49]:
!pip install mne
!wget https://raw.githubusercontent.com/adasegroup/NEUROML2020/seminar1/seminar1/train.csv
!wget https://raw.githubusercontent.com/adasegroup/NEUROML2020/seminar1/seminar1/test.csv

--2025-12-15 02:01:22--  https://raw.githubusercontent.com/adasegroup/NEUROML2020/seminar1/seminar1/train.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 5381138 (5.1M) [text/plain]
Saving to: ‘train.csv.2’

train.csv.2         100%[===================>]   5.13M  --.-KB/s    in 0.08s   

2025-12-15 02:01:22 (64.4 MB/s) - ‘train.csv.2’ saved [5381138/5381138]

--2025-12-15 02:01:22--  https://raw.githubusercontent.com/adasegroup/NEUROML2020/seminar1/seminar1/test.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 3567877 (3.4M

In [50]:
!pip install lightautoml

In [70]:
import pandas as pd
import numpy as np
import mne

from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.impute import SimpleImputer
from sklearn.feature_selection import SelectFromModel
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score, f1_score

from lightgbm import LGBMClassifier
from catboost import CatBoostClassifier

from lightautoml.automl.presets.tabular_presets import TabularAutoML
from lightautoml.tasks import Task

from src.dataset import get_target, calc_features

delimiter = "\n" + "-"*10

In [52]:
df_train = pd.read_csv('train.csv')
df_test = pd.read_csv('test.csv')
df_train.head()

time  condition  epoch         C3         Cz         C4        Fp1  \
0     0          1      0  -5.885714  -2.533107   9.866895   7.962973   
1     1          1      0  -7.999715 -16.916729 -11.924855  17.955477   
2     2          1      0  -6.727283 -15.979567 -11.114195  17.183478   
3     3          1      0   6.819390  -0.204905  10.090124  20.265222   
4     4          1      0  13.129486  -5.817193   5.040633  19.462210   

        Fp2         F7         F3  ...         F8         T7        T8  \
0  5.694433  23.638605  27.899784  ...  23.588723  12.178548  0.685809   
1  8.526994  56.635981  28.508435  ...   1.045533  14.656061 -4.119778   
2  4.497028  43.914130  10.079754  ... -14.741630  14.793562 -6.624813   
3  7.843006  36.250611  13.291199  ...  -7.135541  21.723418 -2.276825   
4  9.634234  42.311729  20.641012  ...   0.015602  19.703190  3.739076   

         P7         P3         Pz         P4         P8         O1        O2  
0 -4.887397 -10.646985 -14.735646  -8.729323  -0.562578 -17.055458 -3.616732  
1 -4.632381 -17.980657 -23.456960 -12.960684  -9.639784 -20.233549 -1.229811  
2 -3.402757 -10.269473 -18.736144  -3.579046  -0.045658 -14.089755  2.453398  
3  2.066859   4.325365  -2.803322   8.835114  13.878945  -5.772410  7.657873  
4  2.714350   3.251047  -3.631448   3.212956   4.635574  -5.045448  5.086024  

[5 rows x 22 columns]

In [53]:
non_eeg = ['time', 'condition', 'epoch', 'target']
channels = [ch for ch in df_train.columns if ch not in non_eeg]

In [54]:
X = get_target(df_train)
X = X.merge(calc_features(df_train, channels), on='epoch')
y = X['condition'].apply(lambda x: 0 if x == 1 else 1)

X = X.drop(['epoch', 'condition'], axis=1)
y = y.astype(int)
y.name = 'target'
X = X.reset_index(drop=True)
y = y.reset_index(drop=True)

## 3. Моделирование (максимум 3.5 балла)
[0.25] Обоснование стратегии разделения данных (train-test split)
Особое внимание уделить предотвращению утечки данных

[0.25] LAMA бейзлайн:
- Минимум 2 различные конфигурации
- Выбор лучшего решения

[3.0] Собственное решение (если не удалось побить LLama baseline: 3 x 1.0 балл за различные пайплайны/попытки):
- Выбор модели
- Построение пайплайна (препроцессинг, обработка пропусков, генерация признаков, отбор признаков, финальная модель/ансамбль)
- Оптимизация гиперпараметров

In [55]:
# Cтратегии разделения данных:
# В соответствием с выводами из файла EDA (part 1) делаем корректный train-test split по данным, предварительно подготовленным по эпохам
npr = np.random.RandomState(42)
epoch_list = X.index
train_idx, test_idx = train_test_split(epoch_list, test_size=0.33, random_state=42) ### делаем разбиение по эпохам как независимым объектам, чтобы избежать утечки между зависимыми наблюдениями, поскольку эпохи коррелированы внутри себя

X_train = X.iloc[train_idx]
X_test = X.iloc[test_idx]
y_train = y.iloc[train_idx]
y_test = y.iloc[test_idx]


In [56]:
# Подготовим данные для конфигураций
roles = {'target': 'target'}
task = Task('binary')

train_LAMA = X_train.copy()
train_LAMA['target'] = y_train.values

test_LAMA = X_test.copy()
test_LAMA['target'] = y_test.values

#### Конфигурация 1 - Классический LAMA (default)
Стандартная конфигурация - подбор лучших моделей (LGB, CatBoost и др.) с автотюнингом. Используем пресеты "по умолчанию".

In [74]:
automl1 = TabularAutoML(
    task=task,
    timeout=600,
    cpu_limit=2,
)
oof_pred1 = automl1.fit_predict(train_LAMA, roles=roles, verbose=2)
test_pred1 = automl1.predict(test_LAMA)

auc1 = roc_auc_score(y_test, test_pred1.data[:, 0])
f1_1 = f1_score(y_test, test_pred1.data[:, 0] > 0.5)
print(f"[LAMA Default] ROC-AUC: {auc1:.4f} F1: {f1_1:.4f}")

[02:26:14] Stdout logging level is INFO2.


INFO:lightautoml.automl.presets.base:Stdout logging level is INFO2.


[02:26:14] Task: binary



INFO:lightautoml.automl.presets.base:Task: binary



[02:26:14] Start automl preset with listed constraints:


INFO:lightautoml.automl.presets.base:Start automl preset with listed constraints:


[02:26:14] - time: 600.00 seconds


INFO:lightautoml.automl.presets.base:- time: 600.00 seconds


[02:26:14] - CPU: 2 cores


INFO:lightautoml.automl.presets.base:- CPU: 2 cores


[02:26:14] - memory: 16 GB



INFO:lightautoml.automl.presets.base:- memory: 16 GB



[02:26:14] Train data shape: (120, 20)



INFO:lightautoml.reader.base:Train data shape: (120, 20)

INFO3:lightautoml.reader.base:Feats was rejected during automatic roles guess: []


[02:26:28] Layer 1 train process start. Time left 585.72 secs


INFO:lightautoml.automl.base:Layer 1 train process start. Time left 585.72 secs


[02:26:28] Start fitting Lvl_0_Pipe_0_Mod_0_LinearL2 ...


INFO:lightautoml.ml_algo.base:Start fitting Lvl_0_Pipe_0_Mod_0_LinearL2 ...
DEBUG:lightautoml.ml_algo.base:Training params: {'tol': 1e-06, 'max_iter': 100, 'cs': [1e-05, 5e-05, 0.0001, 0.0005, 0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1, 5, 10, 50, 100, 500, 1000, 5000, 10000, 50000, 100000], 'early_stopping': 2, 'categorical_idx': [0, 1], 'embed_sizes': array([11, 11], dtype=int32), 'data_size': 27}


[02:26:28] ===== Start working with fold 0 for Lvl_0_Pipe_0_Mod_0_LinearL2 =====


INFO2:lightautoml.ml_algo.base:===== Start working with fold 0 for Lvl_0_Pipe_0_Mod_0_LinearL2 =====
INFO3:lightautoml.ml_algo.torch_based.linear_model:Linear model: C = 1e-05 score = 0.5555555555555556
INFO3:lightautoml.ml_algo.torch_based.linear_model:Linear model: C = 5e-05 score = 0.5625
INFO3:lightautoml.ml_algo.torch_based.linear_model:Linear model: C = 0.0001 score = 0.5625
INFO3:lightautoml.ml_algo.torch_based.linear_model:Linear model: C = 0.0005 score = 0.5694444444444444
INFO3:lightautoml.ml_algo.torch_based.linear_model:Linear model: C = 0.001 score = 0.5763888888888888
INFO3:lightautoml.ml_algo.torch_based.linear_model:Linear model: C = 0.005 score = 0.5694444444444444
INFO3:lightautoml.ml_algo.torch_based.linear_model:Linear model: C = 0.01 score = 0.5694444444444444


[02:26:28] ===== Start working with fold 1 for Lvl_0_Pipe_0_Mod_0_LinearL2 =====


INFO2:lightautoml.ml_algo.base:===== Start working with fold 1 for Lvl_0_Pipe_0_Mod_0_LinearL2 =====
INFO3:lightautoml.ml_algo.torch_based.linear_model:Linear model: C = 1e-05 score = 0.7062937062937064
INFO3:lightautoml.ml_algo.torch_based.linear_model:Linear model: C = 5e-05 score = 0.7062937062937064
INFO3:lightautoml.ml_algo.torch_based.linear_model:Linear model: C = 0.0001 score = 0.7062937062937064


[02:26:29] ===== Start working with fold 2 for Lvl_0_Pipe_0_Mod_0_LinearL2 =====


INFO2:lightautoml.ml_algo.base:===== Start working with fold 2 for Lvl_0_Pipe_0_Mod_0_LinearL2 =====
INFO3:lightautoml.ml_algo.torch_based.linear_model:Linear model: C = 1e-05 score = 0.8041958041958042
INFO3:lightautoml.ml_algo.torch_based.linear_model:Linear model: C = 5e-05 score = 0.8041958041958042
INFO3:lightautoml.ml_algo.torch_based.linear_model:Linear model: C = 0.0001 score = 0.8041958041958042


[02:26:29] ===== Start working with fold 3 for Lvl_0_Pipe_0_Mod_0_LinearL2 =====


INFO2:lightautoml.ml_algo.base:===== Start working with fold 3 for Lvl_0_Pipe_0_Mod_0_LinearL2 =====
INFO3:lightautoml.ml_algo.torch_based.linear_model:Linear model: C = 1e-05 score = 0.4755244755244755
INFO3:lightautoml.ml_algo.torch_based.linear_model:Linear model: C = 5e-05 score = 0.4755244755244755
INFO3:lightautoml.ml_algo.torch_based.linear_model:Linear model: C = 0.0001 score = 0.4755244755244755


[02:26:29] ===== Start working with fold 4 for Lvl_0_Pipe_0_Mod_0_LinearL2 =====


INFO2:lightautoml.ml_algo.base:===== Start working with fold 4 for Lvl_0_Pipe_0_Mod_0_LinearL2 =====
INFO3:lightautoml.ml_algo.torch_based.linear_model:Linear model: C = 1e-05 score = 0.951048951048951
INFO3:lightautoml.ml_algo.torch_based.linear_model:Linear model: C = 5e-05 score = 0.951048951048951
INFO3:lightautoml.ml_algo.torch_based.linear_model:Linear model: C = 0.0001 score = 0.951048951048951


[02:26:29] Fitting Lvl_0_Pipe_0_Mod_0_LinearL2 finished. score = 0.5680803571428572


INFO:lightautoml.ml_algo.base:Fitting Lvl_0_Pipe_0_Mod_0_LinearL2 finished. score = 0.5680803571428572


[02:26:29] Lvl_0_Pipe_0_Mod_0_LinearL2 fitting and predicting completed


INFO:lightautoml.ml_algo.base:Lvl_0_Pipe_0_Mod_0_LinearL2 fitting and predicting completed


[02:26:29] Time left 584.66 secs



INFO:lightautoml.automl.base:Time left 584.66 secs

INFO3:lightautoml.ml_algo.boost_lgbm:Training until validation scores don't improve for 200 rounds
INFO3:lightautoml.ml_algo.boost_lgbm:[100]	valid's auc: 0.520833
INFO3:lightautoml.ml_algo.boost_lgbm:[200]	valid's auc: 0.534722
INFO3:lightautoml.ml_algo.boost_lgbm:Early stopping, best iteration is:
[51]	valid's auc: 0.569444


[02:26:29] Selector_LightGBM fitting and predicting completed


INFO:lightautoml.ml_algo.base:Selector_LightGBM fitting and predicting completed


[02:26:29] Start fitting Lvl_0_Pipe_1_Mod_0_LightGBM ...


INFO:lightautoml.ml_algo.base:Start fitting Lvl_0_Pipe_1_Mod_0_LightGBM ...
DEBUG:lightautoml.ml_algo.base:Training params: {'task': 'train', 'learning_rate': 0.01, 'num_leaves': 16, 'feature_fraction': 0.7, 'bagging_fraction': 0.7, 'bagging_freq': 1, 'max_depth': -1, 'verbosity': -1, 'reg_alpha': 1, 'reg_lambda': 0.0, 'min_split_gain': 0.0, 'zero_as_missing': False, 'num_threads': 2, 'max_bin': 255, 'min_data_in_bin': 3, 'num_trees': 3000, 'early_stopping_rounds': 200, 'random_state': 42, 'verbose_eval': 100}


[02:26:29] ===== Start working with fold 0 for Lvl_0_Pipe_1_Mod_0_LightGBM =====


INFO2:lightautoml.ml_algo.base:===== Start working with fold 0 for Lvl_0_Pipe_1_Mod_0_LightGBM =====
INFO3:lightautoml.ml_algo.boost_lgbm:Training until validation scores don't improve for 200 rounds
INFO3:lightautoml.ml_algo.boost_lgbm:[100]	valid's auc: 0.5625
INFO3:lightautoml.ml_algo.boost_lgbm:[200]	valid's auc: 0.5625
INFO3:lightautoml.ml_algo.boost_lgbm:[300]	valid's auc: 0.548611
INFO3:lightautoml.ml_algo.boost_lgbm:Early stopping, best iteration is:
[198]	valid's auc: 0.576389


[02:26:29] ===== Start working with fold 1 for Lvl_0_Pipe_1_Mod_0_LightGBM =====


INFO2:lightautoml.ml_algo.base:===== Start working with fold 1 for Lvl_0_Pipe_1_Mod_0_LightGBM =====
INFO3:lightautoml.ml_algo.boost_lgbm:Training until validation scores don't improve for 200 rounds
INFO3:lightautoml.ml_algo.boost_lgbm:[100]	valid's auc: 0.727273
INFO3:lightautoml.ml_algo.boost_lgbm:[200]	valid's auc: 0.734266
INFO3:lightautoml.ml_algo.boost_lgbm:Early stopping, best iteration is:
[2]	valid's auc: 0.800699


[02:26:29] ===== Start working with fold 2 for Lvl_0_Pipe_1_Mod_0_LightGBM =====


INFO2:lightautoml.ml_algo.base:===== Start working with fold 2 for Lvl_0_Pipe_1_Mod_0_LightGBM =====
INFO3:lightautoml.ml_algo.boost_lgbm:Training until validation scores don't improve for 200 rounds
INFO3:lightautoml.ml_algo.boost_lgbm:[100]	valid's auc: 0.797203
INFO3:lightautoml.ml_algo.boost_lgbm:[200]	valid's auc: 0.811189
INFO3:lightautoml.ml_algo.boost_lgbm:Early stopping, best iteration is:
[89]	valid's auc: 0.825175


[02:26:29] ===== Start working with fold 3 for Lvl_0_Pipe_1_Mod_0_LightGBM =====


INFO2:lightautoml.ml_algo.base:===== Start working with fold 3 for Lvl_0_Pipe_1_Mod_0_LightGBM =====
INFO3:lightautoml.ml_algo.boost_lgbm:Training until validation scores don't improve for 200 rounds
INFO3:lightautoml.ml_algo.boost_lgbm:[100]	valid's auc: 0.611888
INFO3:lightautoml.ml_algo.boost_lgbm:[200]	valid's auc: 0.625874
INFO3:lightautoml.ml_algo.boost_lgbm:Early stopping, best iteration is:
[4]	valid's auc: 0.65035


[02:26:29] ===== Start working with fold 4 for Lvl_0_Pipe_1_Mod_0_LightGBM =====


INFO2:lightautoml.ml_algo.base:===== Start working with fold 4 for Lvl_0_Pipe_1_Mod_0_LightGBM =====
INFO3:lightautoml.ml_algo.boost_lgbm:Training until validation scores don't improve for 200 rounds
INFO3:lightautoml.ml_algo.boost_lgbm:[100]	valid's auc: 0.86014
INFO3:lightautoml.ml_algo.boost_lgbm:[200]	valid's auc: 0.818182
INFO3:lightautoml.ml_algo.boost_lgbm:[300]	valid's auc: 0.818182
INFO3:lightautoml.ml_algo.boost_lgbm:Early stopping, best iteration is:
[103]	valid's auc: 0.867133


[02:26:29] Fitting Lvl_0_Pipe_1_Mod_0_LightGBM finished. score = 0.6541573660714286


INFO:lightautoml.ml_algo.base:Fitting Lvl_0_Pipe_1_Mod_0_LightGBM finished. score = 0.6541573660714286


[02:26:29] Lvl_0_Pipe_1_Mod_0_LightGBM fitting and predicting completed


INFO:lightautoml.ml_algo.base:Lvl_0_Pipe_1_Mod_0_LightGBM fitting and predicting completed


[02:26:29] Start hyperparameters optimization for Lvl_0_Pipe_1_Mod_1_Tuned_LightGBM ... Time budget is 130.11 secs


INFO:lightautoml.ml_algo.tuning.optuna:Start hyperparameters optimization for Lvl_0_Pipe_1_Mod_1_Tuned_LightGBM ... Time budget is 130.11 secs
Optimization Progress:   0%|          | 0/101 [00:00<?, ?it/s]INFO:optuna.storages._in_memory:A new study created in memory with name: no-name-e9788c11-99bb-41fc-9790-cccdf175966c
INFO3:lightautoml.ml_algo.boost_lgbm:Training until validation scores don't improve for 200 rounds
INFO3:lightautoml.ml_algo.boost_lgbm:[100]	valid's auc: 0.513889
INFO3:lightautoml.ml_algo.boost_lgbm:[200]	valid's auc: 0.5
INFO3:lightautoml.ml_algo.boost_lgbm:Early stopping, best iteration is:
[62]	valid's auc: 0.527778
INFO:optuna.study.study:Trial 0 finished with value: 0.5277777777777779 and parameters: {'feature_fraction': 0.6872700594236812, 'num_leaves': 244, 'bagging_fraction': 0.8659969709057025, 'min_sum_hessian_in_leaf': 0.24810409748678125, 'reg_alpha': 2.5361081166471375e-07, 'reg_lambda': 2.5348407664333426e-07}. Best is trial 0 with value: 0.527777777777

[02:26:43] Hyperparameters optimization for Lvl_0_Pipe_1_Mod_1_Tuned_LightGBM completed



INFO:lightautoml.ml_algo.tuning.optuna:Hyperparameters optimization for Lvl_0_Pipe_1_Mod_1_Tuned_LightGBM completed


[02:26:43] The set of hyperparameters {'feature_fraction': 0.9713386595732587, 'num_leaves': 28, 'bagging_fraction': 0.5263780042015709, 'min_sum_hessian_in_leaf': 0.0014715560349285325, 'reg_alpha': 0.005765667774595165, 'reg_lambda': 3.843066912688489e-07}
 achieve 0.6354 auc


INFO2:lightautoml.ml_algo.tuning.optuna:The set of hyperparameters {'feature_fraction': 0.9713386595732587, 'num_leaves': 28, 'bagging_fraction': 0.5263780042015709, 'min_sum_hessian_in_leaf': 0.0014715560349285325, 'reg_alpha': 0.005765667774595165, 'reg_lambda': 3.843066912688489e-07}
 achieve 0.6354 auc


[02:26:43] Start fitting Lvl_0_Pipe_1_Mod_1_Tuned_LightGBM ...


INFO:lightautoml.ml_algo.base:Start fitting Lvl_0_Pipe_1_Mod_1_Tuned_LightGBM ...
DEBUG:lightautoml.ml_algo.base:Training params: {'task': 'train', 'learning_rate': 0.05, 'num_leaves': 28, 'feature_fraction': 0.9713386595732587, 'bagging_fraction': 0.5263780042015709, 'bagging_freq': 1, 'max_depth': -1, 'verbosity': -1, 'reg_alpha': 0.005765667774595165, 'reg_lambda': 3.843066912688489e-07, 'min_split_gain': 0.0, 'zero_as_missing': False, 'num_threads': 2, 'max_bin': 255, 'min_data_in_bin': 3, 'num_trees': 3000, 'early_stopping_rounds': 100, 'random_state': 42, 'verbose_eval': 100, 'min_sum_hessian_in_leaf': 0.0014715560349285325}


[02:26:43] ===== Start working with fold 0 for Lvl_0_Pipe_1_Mod_1_Tuned_LightGBM =====


INFO2:lightautoml.ml_algo.base:===== Start working with fold 0 for Lvl_0_Pipe_1_Mod_1_Tuned_LightGBM =====
INFO3:lightautoml.ml_algo.boost_lgbm:Training until validation scores don't improve for 100 rounds
INFO3:lightautoml.ml_algo.boost_lgbm:[100]	valid's auc: 0.576389
INFO3:lightautoml.ml_algo.boost_lgbm:Early stopping, best iteration is:
[5]	valid's auc: 0.614583


[02:26:43] ===== Start working with fold 1 for Lvl_0_Pipe_1_Mod_1_Tuned_LightGBM =====


INFO2:lightautoml.ml_algo.base:===== Start working with fold 1 for Lvl_0_Pipe_1_Mod_1_Tuned_LightGBM =====
INFO3:lightautoml.ml_algo.boost_lgbm:Training until validation scores don't improve for 100 rounds
INFO3:lightautoml.ml_algo.boost_lgbm:[100]	valid's auc: 0.566434
INFO3:lightautoml.ml_algo.boost_lgbm:Early stopping, best iteration is:
[3]	valid's auc: 0.706294


[02:26:44] ===== Start working with fold 2 for Lvl_0_Pipe_1_Mod_1_Tuned_LightGBM =====


INFO2:lightautoml.ml_algo.base:===== Start working with fold 2 for Lvl_0_Pipe_1_Mod_1_Tuned_LightGBM =====
INFO3:lightautoml.ml_algo.boost_lgbm:Training until validation scores don't improve for 100 rounds
INFO3:lightautoml.ml_algo.boost_lgbm:[100]	valid's auc: 0.727273
INFO3:lightautoml.ml_algo.boost_lgbm:Early stopping, best iteration is:
[20]	valid's auc: 0.863636


[02:26:44] ===== Start working with fold 3 for Lvl_0_Pipe_1_Mod_1_Tuned_LightGBM =====


INFO2:lightautoml.ml_algo.base:===== Start working with fold 3 for Lvl_0_Pipe_1_Mod_1_Tuned_LightGBM =====
INFO3:lightautoml.ml_algo.boost_lgbm:Training until validation scores don't improve for 100 rounds
INFO3:lightautoml.ml_algo.boost_lgbm:[100]	valid's auc: 0.611888
INFO3:lightautoml.ml_algo.boost_lgbm:Early stopping, best iteration is:
[23]	valid's auc: 0.65035


[02:26:44] ===== Start working with fold 4 for Lvl_0_Pipe_1_Mod_1_Tuned_LightGBM =====


INFO2:lightautoml.ml_algo.base:===== Start working with fold 4 for Lvl_0_Pipe_1_Mod_1_Tuned_LightGBM =====
INFO3:lightautoml.ml_algo.boost_lgbm:Training until validation scores don't improve for 100 rounds
INFO3:lightautoml.ml_algo.boost_lgbm:[100]	valid's auc: 0.867133
INFO3:lightautoml.ml_algo.boost_lgbm:Early stopping, best iteration is:
[6]	valid's auc: 0.93007


[02:26:44] Fitting Lvl_0_Pipe_1_Mod_1_Tuned_LightGBM finished. score = 0.6961495535714286


INFO:lightautoml.ml_algo.base:Fitting Lvl_0_Pipe_1_Mod_1_Tuned_LightGBM finished. score = 0.6961495535714286


[02:26:44] Lvl_0_Pipe_1_Mod_1_Tuned_LightGBM fitting and predicting completed


INFO:lightautoml.ml_algo.base:Lvl_0_Pipe_1_Mod_1_Tuned_LightGBM fitting and predicting completed


[02:26:44] Start fitting Lvl_0_Pipe_1_Mod_2_CatBoost ...


INFO:lightautoml.ml_algo.base:Start fitting Lvl_0_Pipe_1_Mod_2_CatBoost ...
DEBUG:lightautoml.ml_algo.base:Training params: {'task_type': 'CPU', 'thread_count': 2, 'random_seed': 42, 'num_trees': 500, 'learning_rate': 0.02, 'l2_leaf_reg': 0.01, 'bootstrap_type': 'Bernoulli', 'grow_policy': 'SymmetricTree', 'max_depth': 5, 'min_data_in_leaf': 1, 'one_hot_max_size': 10, 'fold_permutation_block': 1, 'boosting_type': 'Plain', 'boost_from_average': True, 'od_type': 'Iter', 'od_wait': 100, 'max_bin': 32, 'feature_border_type': 'GreedyLogSum', 'nan_mode': 'Min', 'verbose': 100, 'allow_writing_files': False, 'verbose_eval': 100}


[02:26:44] ===== Start working with fold 0 for Lvl_0_Pipe_1_Mod_2_CatBoost =====


INFO2:lightautoml.ml_algo.base:===== Start working with fold 0 for Lvl_0_Pipe_1_Mod_2_CatBoost =====
INFO3:lightautoml.ml_algo.boost_cb:0:	test: 0.4409722	best: 0.4409722 (0)	total: 4.32ms	remaining: 2.16s
INFO3:lightautoml.ml_algo.boost_cb:100:	test: 0.5347222	best: 0.5347222 (84)	total: 115ms	remaining: 453ms
INFO3:lightautoml.ml_algo.boost_cb:Stopped by overfitting detector  (100 iterations wait)
INFO3:lightautoml.ml_algo.boost_cb:bestTest = 0.5347222222
INFO3:lightautoml.ml_algo.boost_cb:bestIteration = 84
INFO3:lightautoml.ml_algo.boost_cb:Shrink model to first 85 iterations.


[02:26:44] ===== Start working with fold 1 for Lvl_0_Pipe_1_Mod_2_CatBoost =====


INFO2:lightautoml.ml_algo.base:===== Start working with fold 1 for Lvl_0_Pipe_1_Mod_2_CatBoost =====
INFO3:lightautoml.ml_algo.boost_cb:0:	test: 0.5174825	best: 0.5174825 (0)	total: 1.18ms	remaining: 587ms
INFO3:lightautoml.ml_algo.boost_cb:100:	test: 0.6713287	best: 0.7062937 (49)	total: 94.6ms	remaining: 374ms
INFO3:lightautoml.ml_algo.boost_cb:Stopped by overfitting detector  (100 iterations wait)
INFO3:lightautoml.ml_algo.boost_cb:bestTest = 0.7062937063
INFO3:lightautoml.ml_algo.boost_cb:bestIteration = 49
INFO3:lightautoml.ml_algo.boost_cb:Shrink model to first 50 iterations.


[02:26:44] ===== Start working with fold 2 for Lvl_0_Pipe_1_Mod_2_CatBoost =====


INFO2:lightautoml.ml_algo.base:===== Start working with fold 2 for Lvl_0_Pipe_1_Mod_2_CatBoost =====
INFO3:lightautoml.ml_algo.boost_cb:0:	test: 0.7762238	best: 0.7762238 (0)	total: 999us	remaining: 499ms
INFO3:lightautoml.ml_algo.boost_cb:100:	test: 0.8111888	best: 0.8951049 (8)	total: 95.3ms	remaining: 377ms
INFO3:lightautoml.ml_algo.boost_cb:Stopped by overfitting detector  (100 iterations wait)
INFO3:lightautoml.ml_algo.boost_cb:bestTest = 0.8951048951
INFO3:lightautoml.ml_algo.boost_cb:bestIteration = 8
INFO3:lightautoml.ml_algo.boost_cb:Shrink model to first 9 iterations.


[02:26:44] ===== Start working with fold 3 for Lvl_0_Pipe_1_Mod_2_CatBoost =====


INFO2:lightautoml.ml_algo.base:===== Start working with fold 3 for Lvl_0_Pipe_1_Mod_2_CatBoost =====
INFO3:lightautoml.ml_algo.boost_cb:0:	test: 0.6503497	best: 0.6503497 (0)	total: 1ms	remaining: 502ms
INFO3:lightautoml.ml_algo.boost_cb:100:	test: 0.6083916	best: 0.6608392 (2)	total: 87ms	remaining: 344ms
INFO3:lightautoml.ml_algo.boost_cb:Stopped by overfitting detector  (100 iterations wait)
INFO3:lightautoml.ml_algo.boost_cb:bestTest = 0.6608391608
INFO3:lightautoml.ml_algo.boost_cb:bestIteration = 2
INFO3:lightautoml.ml_algo.boost_cb:Shrink model to first 3 iterations.


[02:26:44] ===== Start working with fold 4 for Lvl_0_Pipe_1_Mod_2_CatBoost =====


INFO2:lightautoml.ml_algo.base:===== Start working with fold 4 for Lvl_0_Pipe_1_Mod_2_CatBoost =====
INFO3:lightautoml.ml_algo.boost_cb:0:	test: 0.5944056	best: 0.5944056 (0)	total: 1.35ms	remaining: 676ms
INFO3:lightautoml.ml_algo.boost_cb:100:	test: 0.8881119	best: 0.8951049 (10)	total: 110ms	remaining: 435ms
INFO3:lightautoml.ml_algo.boost_cb:200:	test: 0.8321678	best: 0.9020979 (101)	total: 199ms	remaining: 297ms
INFO3:lightautoml.ml_algo.boost_cb:Stopped by overfitting detector  (100 iterations wait)
INFO3:lightautoml.ml_algo.boost_cb:bestTest = 0.9020979021
INFO3:lightautoml.ml_algo.boost_cb:bestIteration = 101
INFO3:lightautoml.ml_algo.boost_cb:Shrink model to first 102 iterations.


[02:26:45] Fitting Lvl_0_Pipe_1_Mod_2_CatBoost finished. score = 0.6544363839285715


INFO:lightautoml.ml_algo.base:Fitting Lvl_0_Pipe_1_Mod_2_CatBoost finished. score = 0.6544363839285715


[02:26:45] Lvl_0_Pipe_1_Mod_2_CatBoost fitting and predicting completed


INFO:lightautoml.ml_algo.base:Lvl_0_Pipe_1_Mod_2_CatBoost fitting and predicting completed


[02:26:45] Start hyperparameters optimization for Lvl_0_Pipe_1_Mod_3_Tuned_CatBoost ... Time budget is 300.00 secs


INFO:lightautoml.ml_algo.tuning.optuna:Start hyperparameters optimization for Lvl_0_Pipe_1_Mod_3_Tuned_CatBoost ... Time budget is 300.00 secs
Optimization Progress:   0%|          | 0/101 [00:00<?, ?it/s]INFO:optuna.storages._in_memory:A new study created in memory with name: no-name-9e9c6c5d-5e13-4aa4-bad2-93ea4ebb1b9f
INFO3:lightautoml.ml_algo.boost_cb:0:	test: 0.4548611	best: 0.4548611 (0)	total: 3.95ms	remaining: 1.97s
INFO3:lightautoml.ml_algo.boost_cb:100:	test: 0.5347222	best: 0.5486111 (64)	total: 90.5ms	remaining: 358ms
INFO3:lightautoml.ml_algo.boost_cb:Stopped by overfitting detector  (100 iterations wait)
INFO3:lightautoml.ml_algo.boost_cb:bestTest = 0.5486111111
INFO3:lightautoml.ml_algo.boost_cb:bestIteration = 64
INFO3:lightautoml.ml_algo.boost_cb:Shrink model to first 65 iterations.
INFO:optuna.study.study:Trial 0 finished with value: 0.5486111111111112 and parameters: {'max_depth': 4, 'l2_leaf_reg': 3.6010467344475403, 'min_data_in_leaf': 15}. Best is trial 0 with val

[02:27:22] Hyperparameters optimization for Lvl_0_Pipe_1_Mod_3_Tuned_CatBoost completed



INFO:lightautoml.ml_algo.tuning.optuna:Hyperparameters optimization for Lvl_0_Pipe_1_Mod_3_Tuned_CatBoost completed


[02:27:22] The set of hyperparameters {'max_depth': 4, 'l2_leaf_reg': 5.347108898782771e-06, 'min_data_in_leaf': 12}
 achieve 0.6250 auc


INFO2:lightautoml.ml_algo.tuning.optuna:The set of hyperparameters {'max_depth': 4, 'l2_leaf_reg': 5.347108898782771e-06, 'min_data_in_leaf': 12}
 achieve 0.6250 auc


[02:27:22] Start fitting Lvl_0_Pipe_1_Mod_3_Tuned_CatBoost ...


INFO:lightautoml.ml_algo.base:Start fitting Lvl_0_Pipe_1_Mod_3_Tuned_CatBoost ...
DEBUG:lightautoml.ml_algo.base:Training params: {'task_type': 'CPU', 'thread_count': 2, 'random_seed': 42, 'num_trees': 3000, 'learning_rate': 0.03, 'l2_leaf_reg': 5.347108898782771e-06, 'bootstrap_type': 'Bernoulli', 'grow_policy': 'SymmetricTree', 'max_depth': 4, 'min_data_in_leaf': 12, 'one_hot_max_size': 10, 'fold_permutation_block': 1, 'boosting_type': 'Plain', 'boost_from_average': True, 'od_type': 'Iter', 'od_wait': 100, 'max_bin': 32, 'feature_border_type': 'GreedyLogSum', 'nan_mode': 'Min', 'verbose': 100, 'allow_writing_files': False, 'verbose_eval': 100}


[02:27:22] ===== Start working with fold 0 for Lvl_0_Pipe_1_Mod_3_Tuned_CatBoost =====


INFO2:lightautoml.ml_algo.base:===== Start working with fold 0 for Lvl_0_Pipe_1_Mod_3_Tuned_CatBoost =====
INFO3:lightautoml.ml_algo.boost_cb:0:	test: 0.4479167	best: 0.4479167 (0)	total: 4.43ms	remaining: 13.3s
INFO3:lightautoml.ml_algo.boost_cb:100:	test: 0.5694444	best: 0.5972222 (77)	total: 214ms	remaining: 6.14s
INFO3:lightautoml.ml_algo.boost_cb:Stopped by overfitting detector  (100 iterations wait)
INFO3:lightautoml.ml_algo.boost_cb:bestTest = 0.5972222222
INFO3:lightautoml.ml_algo.boost_cb:bestIteration = 77
INFO3:lightautoml.ml_algo.boost_cb:Shrink model to first 78 iterations.


[02:27:23] ===== Start working with fold 1 for Lvl_0_Pipe_1_Mod_3_Tuned_CatBoost =====


INFO2:lightautoml.ml_algo.base:===== Start working with fold 1 for Lvl_0_Pipe_1_Mod_3_Tuned_CatBoost =====
INFO3:lightautoml.ml_algo.boost_cb:0:	test: 0.5104895	best: 0.5104895 (0)	total: 927us	remaining: 2.78s
INFO3:lightautoml.ml_algo.boost_cb:100:	test: 0.7202797	best: 0.8741259 (33)	total: 100ms	remaining: 2.88s
INFO3:lightautoml.ml_algo.boost_cb:Stopped by overfitting detector  (100 iterations wait)
INFO3:lightautoml.ml_algo.boost_cb:bestTest = 0.8741258741
INFO3:lightautoml.ml_algo.boost_cb:bestIteration = 33
INFO3:lightautoml.ml_algo.boost_cb:Shrink model to first 34 iterations.


[02:27:23] ===== Start working with fold 2 for Lvl_0_Pipe_1_Mod_3_Tuned_CatBoost =====


INFO2:lightautoml.ml_algo.base:===== Start working with fold 2 for Lvl_0_Pipe_1_Mod_3_Tuned_CatBoost =====
INFO3:lightautoml.ml_algo.boost_cb:0:	test: 0.8006993	best: 0.8006993 (0)	total: 3.48ms	remaining: 10.4s
INFO3:lightautoml.ml_algo.boost_cb:100:	test: 0.7622378	best: 0.8321678 (5)	total: 220ms	remaining: 6.32s
INFO3:lightautoml.ml_algo.boost_cb:Stopped by overfitting detector  (100 iterations wait)
INFO3:lightautoml.ml_algo.boost_cb:bestTest = 0.8321678322
INFO3:lightautoml.ml_algo.boost_cb:bestIteration = 5
INFO3:lightautoml.ml_algo.boost_cb:Shrink model to first 6 iterations.


[02:27:23] ===== Start working with fold 3 for Lvl_0_Pipe_1_Mod_3_Tuned_CatBoost =====


INFO2:lightautoml.ml_algo.base:===== Start working with fold 3 for Lvl_0_Pipe_1_Mod_3_Tuned_CatBoost =====
INFO3:lightautoml.ml_algo.boost_cb:0:	test: 0.6503497	best: 0.6503497 (0)	total: 3.25ms	remaining: 9.74s
INFO3:lightautoml.ml_algo.boost_cb:100:	test: 0.6433566	best: 0.7202797 (3)	total: 210ms	remaining: 6.02s
INFO3:lightautoml.ml_algo.boost_cb:Stopped by overfitting detector  (100 iterations wait)
INFO3:lightautoml.ml_algo.boost_cb:bestTest = 0.7202797203
INFO3:lightautoml.ml_algo.boost_cb:bestIteration = 3
INFO3:lightautoml.ml_algo.boost_cb:Shrink model to first 4 iterations.


[02:27:23] ===== Start working with fold 4 for Lvl_0_Pipe_1_Mod_3_Tuned_CatBoost =====


INFO2:lightautoml.ml_algo.base:===== Start working with fold 4 for Lvl_0_Pipe_1_Mod_3_Tuned_CatBoost =====
INFO3:lightautoml.ml_algo.boost_cb:0:	test: 0.6013986	best: 0.6013986 (0)	total: 1.33ms	remaining: 3.98s
INFO3:lightautoml.ml_algo.boost_cb:100:	test: 0.8461538	best: 0.8741259 (73)	total: 204ms	remaining: 5.87s
INFO3:lightautoml.ml_algo.boost_cb:200:	test: 0.8461538	best: 0.8951049 (169)	total: 325ms	remaining: 4.52s
INFO3:lightautoml.ml_algo.boost_cb:Stopped by overfitting detector  (100 iterations wait)
INFO3:lightautoml.ml_algo.boost_cb:bestTest = 0.8951048951
INFO3:lightautoml.ml_algo.boost_cb:bestIteration = 169
INFO3:lightautoml.ml_algo.boost_cb:Shrink model to first 170 iterations.


[02:27:24] Fitting Lvl_0_Pipe_1_Mod_3_Tuned_CatBoost finished. score = 0.7366071428571428


INFO:lightautoml.ml_algo.base:Fitting Lvl_0_Pipe_1_Mod_3_Tuned_CatBoost finished. score = 0.7366071428571428


[02:27:24] Lvl_0_Pipe_1_Mod_3_Tuned_CatBoost fitting and predicting completed


INFO:lightautoml.ml_algo.base:Lvl_0_Pipe_1_Mod_3_Tuned_CatBoost fitting and predicting completed


[02:27:24] Time left 529.68 secs



INFO:lightautoml.automl.base:Time left 529.68 secs



[02:27:24] Layer 1 training completed.



INFO:lightautoml.automl.base:Layer 1 training completed.



[02:27:24] Blending: optimization starts with equal weights. Score = 0.7123326


INFO:lightautoml.automl.blend:Blending: optimization starts with equal weights. Score = 0.7123326


[02:27:24] Blending: iteration 0: score = 0.7385603, weights = [0.41398153 0.06656854 0.14960875 0.         0.3698412 ]


INFO:lightautoml.automl.blend:Blending: iteration 0: score = 0.7385603, weights = [0.41398153 0.06656854 0.14960875 0.         0.3698412 ]


[02:27:24] Blending: iteration 1: score = 0.7424665, weights = [0.46047238 0.         0.15539142 0.         0.38413626]


INFO:lightautoml.automl.blend:Blending: iteration 1: score = 0.7424665, weights = [0.46047238 0.         0.15539142 0.         0.38413626]


[02:27:24] Blending: iteration 2: score = 0.7430246, weights = [0.5419463  0.         0.09016994 0.         0.36788374]


INFO:lightautoml.automl.blend:Blending: iteration 2: score = 0.7430246, weights = [0.5419463  0.         0.09016994 0.         0.36788374]


[02:27:25] Blending: iteration 3: score = 0.7449777, weights = [0.6163315  0.         0.07662547 0.         0.30704305]


INFO:lightautoml.automl.blend:Blending: iteration 3: score = 0.7449777, weights = [0.6163315  0.         0.07662547 0.         0.30704305]


[02:27:25] Blending: no improvements for score. Terminated.



INFO:lightautoml.automl.blend:Blending: no improvements for score. Terminated.



[02:27:25] Blending: best score = 0.7449777, best weights = [0.6163315  0.         0.07662547 0.         0.30704305]


INFO:lightautoml.automl.blend:Blending: best score = 0.7449777, best weights = [0.6163315  0.         0.07662547 0.         0.30704305]


[02:27:25] Automl preset training completed in 71.23 seconds



INFO:lightautoml.automl.presets.base:Automl preset training completed in 71.23 seconds



[02:27:25] Model description:
Final prediction for new objects (level 0) = 
	 0.61633 * (5 averaged models Lvl_0_Pipe_0_Mod_0_LinearL2) +
	 0.07663 * (5 averaged models Lvl_0_Pipe_1_Mod_1_Tuned_LightGBM) +
	 0.30704 * (5 averaged models Lvl_0_Pipe_1_Mod_3_Tuned_CatBoost) 



INFO:lightautoml.automl.presets.base:Model description:
Final prediction for new objects (level 0) = 
	 0.61633 * (5 averaged models Lvl_0_Pipe_0_Mod_0_LinearL2) +
	 0.07663 * (5 averaged models Lvl_0_Pipe_1_Mod_1_Tuned_LightGBM) +
	 0.30704 * (5 averaged models Lvl_0_Pipe_1_Mod_3_Tuned_CatBoost) 



[LAMA Default] ROC-AUC: 0.7217 F1: 0.6557


### Конфигурация 2 — FAST, только lgbm
Используем только быстрый LightGBM.

In [75]:
automl2 = TabularAutoML(
    task=task,
    timeout=300,
    cpu_limit=2,
    general_params={
        'use_algos': [['lgb']]
    }
)
oof_pred2 = automl2.fit_predict(train_LAMA, roles=roles, verbose=1)
test_pred2 = automl2.predict(test_LAMA)

auc2 = roc_auc_score(y_test, test_pred2.data[:, 0])
f1_2 = f1_score(y_test, test_pred2.data[:, 0] > 0.5)
print(f'[LAMA FAST (lgb only)] ROC-AUC: {auc2:.4f}  F1: {f1_2:.4f}')


[02:27:25] Stdout logging level is INFO.


INFO:lightautoml.automl.presets.base:Stdout logging level is INFO.


[02:27:25] Task: binary



INFO:lightautoml.automl.presets.base:Task: binary



[02:27:25] Start automl preset with listed constraints:


INFO:lightautoml.automl.presets.base:Start automl preset with listed constraints:


[02:27:25] - time: 300.00 seconds


INFO:lightautoml.automl.presets.base:- time: 300.00 seconds


[02:27:25] - CPU: 2 cores


INFO:lightautoml.automl.presets.base:- CPU: 2 cores


[02:27:25] - memory: 16 GB



INFO:lightautoml.automl.presets.base:- memory: 16 GB



[02:27:25] Train data shape: (120, 20)



INFO:lightautoml.reader.base:Train data shape: (120, 20)

INFO3:lightautoml.reader.base:Feats was rejected during automatic roles guess: []


[02:27:26] Layer 1 train process start. Time left 299.46 secs


INFO:lightautoml.automl.base:Layer 1 train process start. Time left 299.46 secs
INFO3:lightautoml.ml_algo.boost_lgbm:Training until validation scores don't improve for 200 rounds
INFO3:lightautoml.ml_algo.boost_lgbm:[100]	valid's auc: 0.520833
INFO3:lightautoml.ml_algo.boost_lgbm:[200]	valid's auc: 0.534722
INFO3:lightautoml.ml_algo.boost_lgbm:Early stopping, best iteration is:
[51]	valid's auc: 0.569444


[02:27:26] Selector_LightGBM fitting and predicting completed


INFO:lightautoml.ml_algo.base:Selector_LightGBM fitting and predicting completed


[02:27:26] Start fitting Lvl_0_Pipe_0_Mod_0_LightGBM ...


INFO:lightautoml.ml_algo.base:Start fitting Lvl_0_Pipe_0_Mod_0_LightGBM ...
DEBUG:lightautoml.ml_algo.base:Training params: {'task': 'train', 'learning_rate': 0.01, 'num_leaves': 16, 'feature_fraction': 0.7, 'bagging_fraction': 0.7, 'bagging_freq': 1, 'max_depth': -1, 'verbosity': -1, 'reg_alpha': 1, 'reg_lambda': 0.0, 'min_split_gain': 0.0, 'zero_as_missing': False, 'num_threads': 2, 'max_bin': 255, 'min_data_in_bin': 3, 'num_trees': 3000, 'early_stopping_rounds': 200, 'random_state': 42, 'verbose_eval': 100}
INFO2:lightautoml.ml_algo.base:===== Start working with fold 0 for Lvl_0_Pipe_0_Mod_0_LightGBM =====
INFO3:lightautoml.ml_algo.boost_lgbm:Training until validation scores don't improve for 200 rounds
INFO3:lightautoml.ml_algo.boost_lgbm:[100]	valid's auc: 0.5625
INFO3:lightautoml.ml_algo.boost_lgbm:[200]	valid's auc: 0.5625
INFO3:lightautoml.ml_algo.boost_lgbm:[300]	valid's auc: 0.548611
INFO3:lightautoml.ml_algo.boost_lgbm:Early stopping, best iteration is:
[198]	valid's auc: 0.

[02:27:27] Fitting Lvl_0_Pipe_0_Mod_0_LightGBM finished. score = 0.6541573660714286


INFO:lightautoml.ml_algo.base:Fitting Lvl_0_Pipe_0_Mod_0_LightGBM finished. score = 0.6541573660714286


[02:27:28] Lvl_0_Pipe_0_Mod_0_LightGBM fitting and predicting completed


INFO:lightautoml.ml_algo.base:Lvl_0_Pipe_0_Mod_0_LightGBM fitting and predicting completed


[02:27:28] Time left 297.68 secs



INFO:lightautoml.automl.base:Time left 297.68 secs



[02:27:28] Layer 1 training completed.



INFO:lightautoml.automl.base:Layer 1 training completed.



[02:27:28] Automl preset training completed in 2.33 seconds



INFO:lightautoml.automl.presets.base:Automl preset training completed in 2.33 seconds



[02:27:28] Model description:
Final prediction for new objects (level 0) = 
	 1.00000 * (5 averaged models Lvl_0_Pipe_0_Mod_0_LightGBM) 



INFO:lightautoml.automl.presets.base:Model description:
Final prediction for new objects (level 0) = 
	 1.00000 * (5 averaged models Lvl_0_Pipe_0_Mod_0_LightGBM) 



[LAMA FAST (lgb only)] ROC-AUC: 0.7138  F1: 0.6552


### Конфигурация 3 - расширенный набор моделей и мощный автотюнинг

Включаем сразу несколько ансамблей (XGB, LGBM с тюнингом, CatBoost с тюнингом, MLP нейросеть) + используем продвинутую кросс-валидацию, автотюнинг с разными алгоритмами

In [76]:
automl3 = TabularAutoML(
    task=task,
    timeout=600,
    cpu_limit=2,
    general_params={
        'use_algos': [['xgb', 'lgb_tuned', 'cb_tuned'], ['mlp']],  # Много моделей
    },
    reader_params={
        'n_jobs': 2,
        'cv': 5,
        'random_state': 84,
        'advanced_roles': True,
    },
    tuning_params={'max_tuning_iter': "auto", 'max_tuning_time': 300},
    selection_params={'mode': 1}
)

oof_pred3 = automl3.fit_predict(train_LAMA, roles=roles, verbose=1)
test_pred3 = automl3.predict(test_LAMA)

auc3 = roc_auc_score(y_test, test_pred3.data[:, 0])
f1_3 = f1_score(y_test, test_pred3.data[:, 0] > 0.5)
print(f'[LAMA Advanced] ROC-AUC: {auc3:.4f}  F1: {f1_3:.4f}')


[02:27:28] Stdout logging level is INFO.


INFO:lightautoml.automl.presets.base:Stdout logging level is INFO.


[02:27:28] Task: binary



INFO:lightautoml.automl.presets.base:Task: binary



[02:27:28] Start automl preset with listed constraints:


INFO:lightautoml.automl.presets.base:Start automl preset with listed constraints:


[02:27:28] - time: 600.00 seconds


INFO:lightautoml.automl.presets.base:- time: 600.00 seconds


[02:27:28] - CPU: 2 cores


INFO:lightautoml.automl.presets.base:- CPU: 2 cores


[02:27:28] - memory: 16 GB



INFO:lightautoml.automl.presets.base:- memory: 16 GB



[02:27:28] Train data shape: (120, 20)



INFO:lightautoml.reader.base:Train data shape: (120, 20)

INFO3:lightautoml.reader.base:Feats was rejected during automatic roles guess: []


[02:27:28] Layer 1 train process start. Time left 599.69 secs


INFO:lightautoml.automl.base:Layer 1 train process start. Time left 599.69 secs
INFO3:lightautoml.ml_algo.boost_lgbm:Training until validation scores don't improve for 200 rounds
INFO3:lightautoml.ml_algo.boost_lgbm:[100]	valid's auc: 0.784722
INFO3:lightautoml.ml_algo.boost_lgbm:[200]	valid's auc: 0.743056
INFO3:lightautoml.ml_algo.boost_lgbm:[300]	valid's auc: 0.743056
INFO3:lightautoml.ml_algo.boost_lgbm:Early stopping, best iteration is:
[100]	valid's auc: 0.784722


[02:27:28] Selector_LightGBM fitting and predicting completed


INFO:lightautoml.ml_algo.base:Selector_LightGBM fitting and predicting completed


[02:27:28] Start hyperparameters optimization for Lvl_0_Pipe_0_Mod_0_Tuned_LightGBM ... Time budget is 83.76 secs


INFO:lightautoml.ml_algo.tuning.optuna:Start hyperparameters optimization for Lvl_0_Pipe_0_Mod_0_Tuned_LightGBM ... Time budget is 83.76 secs
Optimization Progress:   0%|          | 0/100 [00:00<?, ?it/s]INFO:optuna.storages._in_memory:A new study created in memory with name: no-name-c32d1f06-37c0-4b53-97bc-4bb1e40055f4
INFO3:lightautoml.ml_algo.boost_lgbm:Training until validation scores don't improve for 200 rounds
INFO3:lightautoml.ml_algo.boost_lgbm:[100]	valid's auc: 0.770833
INFO3:lightautoml.ml_algo.boost_lgbm:[200]	valid's auc: 0.75
INFO3:lightautoml.ml_algo.boost_lgbm:Early stopping, best iteration is:
[64]	valid's auc: 0.791667
INFO:optuna.study.study:Trial 0 finished with value: 0.7916666666666666 and parameters: {'feature_fraction': 0.6872700594236812, 'num_leaves': 244, 'bagging_fraction': 0.8659969709057025, 'min_sum_hessian_in_leaf': 0.24810409748678125}. Best is trial 0 with value: 0.7916666666666666.
INFO3:lightautoml.ml_algo.tuning.optuna:Trial 1 with hyperparameters 

[02:27:49] Hyperparameters optimization for Lvl_0_Pipe_0_Mod_0_Tuned_LightGBM completed



INFO:lightautoml.ml_algo.tuning.optuna:Hyperparameters optimization for Lvl_0_Pipe_0_Mod_0_Tuned_LightGBM completed
INFO2:lightautoml.ml_algo.tuning.optuna:The set of hyperparameters {'feature_fraction': 0.5102734049121492, 'num_leaves': 129, 'bagging_fraction': 0.9847685553939329, 'min_sum_hessian_in_leaf': 0.689539874561655}
 achieve 0.8785 auc


[02:27:49] Start fitting Lvl_0_Pipe_0_Mod_0_Tuned_LightGBM ...


INFO:lightautoml.ml_algo.base:Start fitting Lvl_0_Pipe_0_Mod_0_Tuned_LightGBM ...
DEBUG:lightautoml.ml_algo.base:Training params: {'task': 'train', 'learning_rate': 0.05, 'num_leaves': 129, 'feature_fraction': 0.5102734049121492, 'bagging_fraction': 0.9847685553939329, 'bagging_freq': 1, 'max_depth': -1, 'verbosity': -1, 'reg_alpha': 1, 'reg_lambda': 0.0, 'min_split_gain': 0.0, 'zero_as_missing': False, 'num_threads': 2, 'max_bin': 255, 'min_data_in_bin': 3, 'num_trees': 3000, 'early_stopping_rounds': 100, 'random_state': 42, 'verbose_eval': 100, 'min_sum_hessian_in_leaf': 0.689539874561655}
INFO2:lightautoml.ml_algo.base:===== Start working with fold 0 for Lvl_0_Pipe_0_Mod_0_Tuned_LightGBM =====
INFO3:lightautoml.ml_algo.boost_lgbm:Training until validation scores don't improve for 100 rounds
INFO3:lightautoml.ml_algo.boost_lgbm:[100]	valid's auc: 0.770833
INFO3:lightautoml.ml_algo.boost_lgbm:Early stopping, best iteration is:
[37]	valid's auc: 0.850694
INFO2:lightautoml.ml_algo.base:

[02:27:52] Fitting Lvl_0_Pipe_0_Mod_0_Tuned_LightGBM finished. score = 0.7660435267857142


INFO:lightautoml.ml_algo.base:Fitting Lvl_0_Pipe_0_Mod_0_Tuned_LightGBM finished. score = 0.7660435267857142


[02:27:52] Lvl_0_Pipe_0_Mod_0_Tuned_LightGBM fitting and predicting completed


INFO:lightautoml.ml_algo.base:Lvl_0_Pipe_0_Mod_0_Tuned_LightGBM fitting and predicting completed


[02:27:52] Start hyperparameters optimization for Lvl_0_Pipe_0_Mod_1_Tuned_CatBoost ... Time budget is 135.49 secs


INFO:lightautoml.ml_algo.tuning.optuna:Start hyperparameters optimization for Lvl_0_Pipe_0_Mod_1_Tuned_CatBoost ... Time budget is 135.49 secs
Optimization Progress:   0%|          | 0/100 [00:00<?, ?it/s]INFO:optuna.storages._in_memory:A new study created in memory with name: no-name-f9920959-5d25-46eb-9ff7-6088fb689a89
INFO3:lightautoml.ml_algo.boost_cb:0:	test: 0.7534722	best: 0.7534722 (0)	total: 1.35ms	remaining: 674ms
INFO3:lightautoml.ml_algo.boost_cb:100:	test: 0.8472222	best: 0.8819444 (12)	total: 280ms	remaining: 1.11s
INFO3:lightautoml.ml_algo.boost_cb:Stopped by overfitting detector  (100 iterations wait)
INFO3:lightautoml.ml_algo.boost_cb:bestTest = 0.8819444444
INFO3:lightautoml.ml_algo.boost_cb:bestIteration = 12
INFO3:lightautoml.ml_algo.boost_cb:Shrink model to first 13 iterations.
INFO:optuna.study.study:Trial 0 finished with value: 0.8819444444444444 and parameters: {'max_depth': 4, 'l2_leaf_reg': 3.6010467344475403, 'min_data_in_leaf': 15}. Best is trial 0 with valu

[02:28:24] Hyperparameters optimization for Lvl_0_Pipe_0_Mod_1_Tuned_CatBoost completed



INFO:lightautoml.ml_algo.tuning.optuna:Hyperparameters optimization for Lvl_0_Pipe_0_Mod_1_Tuned_CatBoost completed
INFO2:lightautoml.ml_algo.tuning.optuna:The set of hyperparameters {'max_depth': 3, 'l2_leaf_reg': 1.4142208298719828e-05, 'min_data_in_leaf': 13}
 achieve 0.9306 auc


[02:28:24] Start fitting Lvl_0_Pipe_0_Mod_1_Tuned_CatBoost ...


INFO:lightautoml.ml_algo.base:Start fitting Lvl_0_Pipe_0_Mod_1_Tuned_CatBoost ...
DEBUG:lightautoml.ml_algo.base:Training params: {'task_type': 'CPU', 'thread_count': 2, 'random_seed': 42, 'num_trees': 3000, 'learning_rate': 0.03, 'l2_leaf_reg': 1.4142208298719828e-05, 'bootstrap_type': 'Bernoulli', 'grow_policy': 'SymmetricTree', 'max_depth': 3, 'min_data_in_leaf': 13, 'one_hot_max_size': 10, 'fold_permutation_block': 1, 'boosting_type': 'Plain', 'boost_from_average': True, 'od_type': 'Iter', 'od_wait': 100, 'max_bin': 32, 'feature_border_type': 'GreedyLogSum', 'nan_mode': 'Min', 'verbose': 100, 'allow_writing_files': False, 'verbose_eval': 100}
INFO2:lightautoml.ml_algo.base:===== Start working with fold 0 for Lvl_0_Pipe_0_Mod_1_Tuned_CatBoost =====
INFO3:lightautoml.ml_algo.boost_cb:0:	test: 0.8506944	best: 0.8506944 (0)	total: 840us	remaining: 2.52s
INFO3:lightautoml.ml_algo.boost_cb:100:	test: 0.7569444	best: 0.8923611 (4)	total: 88.2ms	remaining: 2.53s
INFO3:lightautoml.ml_algo.b

[02:28:24] Fitting Lvl_0_Pipe_0_Mod_1_Tuned_CatBoost finished. score = 0.7657645089285713


INFO:lightautoml.ml_algo.base:Fitting Lvl_0_Pipe_0_Mod_1_Tuned_CatBoost finished. score = 0.7657645089285713


[02:28:24] Lvl_0_Pipe_0_Mod_1_Tuned_CatBoost fitting and predicting completed


INFO:lightautoml.ml_algo.base:Lvl_0_Pipe_0_Mod_1_Tuned_CatBoost fitting and predicting completed


[02:28:24] Start fitting Lvl_0_Pipe_0_Mod_2_XGBoost ...


INFO:lightautoml.ml_algo.base:Start fitting Lvl_0_Pipe_0_Mod_2_XGBoost ...
DEBUG:lightautoml.ml_algo.base:Training params: {'n_estimators': 3000, 'early_stopping_rounds': 100, 'seed': 42, 'verbose_eval': 100, 'nthread': 2}
INFO2:lightautoml.ml_algo.base:===== Start working with fold 0 for Lvl_0_Pipe_0_Mod_2_XGBoost =====
INFO3:lightautoml.ml_algo.boost_xgb:[0]	valid-auc:0.76042
INFO3:lightautoml.ml_algo.boost_xgb:[100]	valid-auc:0.77778
INFO3:lightautoml.ml_algo.boost_xgb:[104]	valid-auc:0.77778
INFO2:lightautoml.ml_algo.base:===== Start working with fold 1 for Lvl_0_Pipe_0_Mod_2_XGBoost =====
INFO3:lightautoml.ml_algo.boost_xgb:[0]	valid-auc:0.75524
INFO3:lightautoml.ml_algo.boost_xgb:[100]	valid-auc:0.82517
INFO3:lightautoml.ml_algo.boost_xgb:[112]	valid-auc:0.82517
INFO2:lightautoml.ml_algo.base:===== Start working with fold 2 for Lvl_0_Pipe_0_Mod_2_XGBoost =====
INFO3:lightautoml.ml_algo.boost_xgb:[0]	valid-auc:0.81469
INFO3:lightautoml.ml_algo.boost_xgb:[99]	valid-auc:0.67832
INFO

[02:28:25] Fitting Lvl_0_Pipe_0_Mod_2_XGBoost finished. score = 0.7287946428571428


INFO:lightautoml.ml_algo.base:Fitting Lvl_0_Pipe_0_Mod_2_XGBoost finished. score = 0.7287946428571428


[02:28:25] Lvl_0_Pipe_0_Mod_2_XGBoost fitting and predicting completed


INFO:lightautoml.ml_algo.base:Lvl_0_Pipe_0_Mod_2_XGBoost fitting and predicting completed


[02:28:25] Time left 543.09 secs



INFO:lightautoml.automl.base:Time left 543.09 secs



[02:28:25] Layer 1 training completed.



INFO:lightautoml.automl.base:Layer 1 training completed.



[02:28:25] Layer 2 train process start. Time left 543.07 secs


INFO:lightautoml.automl.base:Layer 2 train process start. Time left 543.07 secs
DEBUG:lightautoml.ml_algo.dl_model:number of text features: 0 
DEBUG:lightautoml.ml_algo.dl_model:number of categorical features: 0 
DEBUG:lightautoml.ml_algo.dl_model:number of continuous features: 22 


[02:28:25] Start fitting Lvl_1_Pipe_0_Mod_0_TorchNN_mlp_0 ...


INFO:lightautoml.ml_algo.base:Start fitting Lvl_1_Pipe_0_Mod_0_TorchNN_mlp_0 ...
DEBUG:lightautoml.ml_algo.base:Training params: {'num_workers': 0, 'pin_memory': False, 'max_length': 256, 'is_snap': False, 'input_bn': False, 'max_emb_size': 256, 'bert_name': None, 'pooling': 'cls', 'device': device(type='cpu'), 'use_cont': True, 'use_cat': True, 'use_text': False, 'lang': 'en', 'deterministic': True, 'multigpu': False, 'random_state': 42, 'model': 'mlp', 'model_with_emb': False, 'path_to_save': None, 'verbose_inside': None, 'verbose': 1, 'n_epochs': 50, 'snap_params': {'k': 3, 'early_stopping': True, 'patience': 10, 'swa': True}, 'bs': 256, 'emb_dropout': 0.1, 'emb_ratio': 3, 'opt': 'Adam', 'opt_params': {'lr': 0.0003, 'weight_decay': 0}, 'sch': 'ReduceLROnPlateau', 'scheduler_params': {'patience': 5, 'factor': 0.5, 'min_lr': 1e-05}, 'loss': None, 'loss_params': {}, 'loss_on_logits': True, 'clip_grad': False, 'clip_grad_params': {}, 'init_bias': True, 'dataset': 'UniversalDataset', 'tu

[02:28:28] Fitting Lvl_1_Pipe_0_Mod_0_TorchNN_mlp_0 finished. score = 0.4866071428571428


INFO:lightautoml.ml_algo.base:Fitting Lvl_1_Pipe_0_Mod_0_TorchNN_mlp_0 finished. score = 0.4866071428571428


[02:28:28] Lvl_1_Pipe_0_Mod_0_TorchNN_mlp_0 fitting and predicting completed


INFO:lightautoml.ml_algo.base:Lvl_1_Pipe_0_Mod_0_TorchNN_mlp_0 fitting and predicting completed


[02:28:28] Time left 539.83 secs



INFO:lightautoml.automl.base:Time left 539.83 secs



[02:28:28] Layer 2 training completed.



INFO:lightautoml.automl.base:Layer 2 training completed.



[02:28:28] Automl preset training completed in 60.17 seconds



INFO:lightautoml.automl.presets.base:Automl preset training completed in 60.17 seconds



[02:28:28] Model description:
Models on level 0:
	 5 averaged models Lvl_0_Pipe_0_Mod_0_Tuned_LightGBM
	 5 averaged models Lvl_0_Pipe_0_Mod_1_Tuned_CatBoost
	 5 averaged models Lvl_0_Pipe_0_Mod_2_XGBoost

Final prediction for new objects (level 1) = 
	 1.00000 * (5 averaged models Lvl_1_Pipe_0_Mod_0_TorchNN_mlp_0) 



INFO:lightautoml.automl.presets.base:Model description:
Models on level 0:
	 5 averaged models Lvl_0_Pipe_0_Mod_0_Tuned_LightGBM
	 5 averaged models Lvl_0_Pipe_0_Mod_1_Tuned_CatBoost
	 5 averaged models Lvl_0_Pipe_0_Mod_2_XGBoost

Final prediction for new objects (level 1) = 
	 1.00000 * (5 averaged models Lvl_1_Pipe_0_Mod_0_TorchNN_mlp_0) 



[LAMA Advanced] ROC-AUC: 0.5000  F1: 0.0000


### конфигурация 4 - с feature selection и ограниченным набором моделей
Только LGBM (c hyperparameter tuning) и линейная регрессия; включён feature selection (оставляем 80% лучших признаков); гиперпараметры подбираются на holdout-валидации.


In [77]:
automl4 = TabularAutoML(
    task=task,
    timeout=600,
    cpu_limit=2,
    general_params={
        'use_algos': [['lgb_tuned', 'linear_l2']]  # Только LGBM + Linear
    },
    selection_params={
        'select_algos': ['gbm', 'linear_l2'],
        'selection_feats_rate': 0.8,
        'max_features_cnt_in_result': None,
    },
    tuning_params={
        'fit_on_holdout': True,
        'max_tuning_iter': 50,
        'max_tuning_time': 200,
    }
)

oof_pred4 = automl4.fit_predict(train_LAMA, roles=roles, verbose=1)
test_pred4 = automl4.predict(test_LAMA)

auc4 = roc_auc_score(y_test, test_pred4.data[:, 0])
f1_4 = f1_score(y_test, test_pred4.data[:, 0] > 0.5)
print(f'[LAMA FS+LGBM/Linear] ROC-AUC: {auc4:.4f}  F1: {f1_4:.4f}')


[02:28:30] Stdout logging level is INFO.


INFO:lightautoml.automl.presets.base:Stdout logging level is INFO.


[02:28:30] Task: binary



INFO:lightautoml.automl.presets.base:Task: binary



[02:28:30] Start automl preset with listed constraints:


INFO:lightautoml.automl.presets.base:Start automl preset with listed constraints:


[02:28:30] - time: 600.00 seconds


INFO:lightautoml.automl.presets.base:- time: 600.00 seconds


[02:28:30] - CPU: 2 cores


INFO:lightautoml.automl.presets.base:- CPU: 2 cores


[02:28:30] - memory: 16 GB



INFO:lightautoml.automl.presets.base:- memory: 16 GB



[02:28:30] Train data shape: (120, 20)



INFO:lightautoml.reader.base:Train data shape: (120, 20)

INFO3:lightautoml.reader.base:Feats was rejected during automatic roles guess: []


[02:28:30] Layer 1 train process start. Time left 599.82 secs


INFO:lightautoml.automl.base:Layer 1 train process start. Time left 599.82 secs
INFO3:lightautoml.ml_algo.boost_lgbm:Training until validation scores don't improve for 200 rounds
INFO3:lightautoml.ml_algo.boost_lgbm:[100]	valid's auc: 0.520833
INFO3:lightautoml.ml_algo.boost_lgbm:[200]	valid's auc: 0.534722
INFO3:lightautoml.ml_algo.boost_lgbm:Early stopping, best iteration is:
[51]	valid's auc: 0.569444


[02:28:30] Selector_LightGBM fitting and predicting completed


INFO:lightautoml.ml_algo.base:Selector_LightGBM fitting and predicting completed


[02:28:30] Start fitting Lvl_0_Pipe_0_Mod_0_LinearL2 ...


INFO:lightautoml.ml_algo.base:Start fitting Lvl_0_Pipe_0_Mod_0_LinearL2 ...
DEBUG:lightautoml.ml_algo.base:Training params: {'tol': 1e-06, 'max_iter': 100, 'cs': [1e-05, 5e-05, 0.0001, 0.0005, 0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1, 5, 10, 50, 100, 500, 1000, 5000, 10000, 50000, 100000], 'early_stopping': 2, 'categorical_idx': [0], 'embed_sizes': array([11], dtype=int32), 'data_size': 17}
INFO2:lightautoml.ml_algo.base:===== Start working with fold 0 for Lvl_0_Pipe_0_Mod_0_LinearL2 =====
INFO3:lightautoml.ml_algo.torch_based.linear_model:Linear model: C = 1e-05 score = 0.5555555555555556
INFO3:lightautoml.ml_algo.torch_based.linear_model:Linear model: C = 5e-05 score = 0.5625
INFO3:lightautoml.ml_algo.torch_based.linear_model:Linear model: C = 0.0001 score = 0.5555555555555556
INFO3:lightautoml.ml_algo.torch_based.linear_model:Linear model: C = 0.0005 score = 0.5555555555555556
INFO2:lightautoml.ml_algo.base:===== Start working with fold 1 for Lvl_0_Pipe_0_Mod_0_LinearL2 =====
INFO3:lig

[02:28:31] Fitting Lvl_0_Pipe_0_Mod_0_LinearL2 finished. score = 0.5680803571428571


INFO:lightautoml.ml_algo.base:Fitting Lvl_0_Pipe_0_Mod_0_LinearL2 finished. score = 0.5680803571428571


[02:28:31] Lvl_0_Pipe_0_Mod_0_LinearL2 fitting and predicting completed


INFO:lightautoml.ml_algo.base:Lvl_0_Pipe_0_Mod_0_LinearL2 fitting and predicting completed


[02:28:31] Time left 598.52 secs



INFO:lightautoml.automl.base:Time left 598.52 secs



[02:28:32] Start hyperparameters optimization for Lvl_0_Pipe_1_Mod_0_Tuned_LightGBM ... Time budget is 200.00 secs


INFO:lightautoml.ml_algo.tuning.optuna:Start hyperparameters optimization for Lvl_0_Pipe_1_Mod_0_Tuned_LightGBM ... Time budget is 200.00 secs
Optimization Progress:   0%|          | 0/50 [00:00<?, ?it/s]INFO:optuna.storages._in_memory:A new study created in memory with name: no-name-c19ac6ac-d1df-4ced-841a-3b2e3a123a43
INFO3:lightautoml.ml_algo.boost_lgbm:Training until validation scores don't improve for 200 rounds
INFO3:lightautoml.ml_algo.boost_lgbm:[100]	valid's auc: 0.520833
INFO3:lightautoml.ml_algo.boost_lgbm:[200]	valid's auc: 0.493056
INFO3:lightautoml.ml_algo.boost_lgbm:Early stopping, best iteration is:
[50]	valid's auc: 0.520833
INFO:optuna.study.study:Trial 0 finished with value: 0.5208333333333334 and parameters: {'feature_fraction': 0.6872700594236812, 'num_leaves': 244, 'bagging_fraction': 0.8659969709057025, 'min_sum_hessian_in_leaf': 0.24810409748678125}. Best is trial 0 with value: 0.5208333333333334.
INFO3:lightautoml.ml_algo.tuning.optuna:Trial 1 with hyperparamet

[02:28:37] Hyperparameters optimization for Lvl_0_Pipe_1_Mod_0_Tuned_LightGBM completed



INFO:lightautoml.ml_algo.tuning.optuna:Hyperparameters optimization for Lvl_0_Pipe_1_Mod_0_Tuned_LightGBM completed
INFO2:lightautoml.ml_algo.tuning.optuna:The set of hyperparameters {'feature_fraction': 0.8124527509966039, 'num_leaves': 144, 'bagging_fraction': 0.5086792017528844, 'min_sum_hessian_in_leaf': 0.17837062520581531}
 achieve 0.6146 auc


[02:28:37] Start fitting Lvl_0_Pipe_1_Mod_0_Tuned_LightGBM ...


INFO:lightautoml.ml_algo.base:Start fitting Lvl_0_Pipe_1_Mod_0_Tuned_LightGBM ...
DEBUG:lightautoml.ml_algo.base:Training params: {'task': 'train', 'learning_rate': 0.05, 'num_leaves': 144, 'feature_fraction': 0.8124527509966039, 'bagging_fraction': 0.5086792017528844, 'bagging_freq': 1, 'max_depth': -1, 'verbosity': -1, 'reg_alpha': 1, 'reg_lambda': 0.0, 'min_split_gain': 0.0, 'zero_as_missing': False, 'num_threads': 2, 'max_bin': 255, 'min_data_in_bin': 3, 'num_trees': 3000, 'early_stopping_rounds': 100, 'random_state': 42, 'verbose_eval': 100, 'min_sum_hessian_in_leaf': 0.17837062520581531}
INFO2:lightautoml.ml_algo.base:===== Start working with fold 0 for Lvl_0_Pipe_1_Mod_0_Tuned_LightGBM =====
INFO3:lightautoml.ml_algo.boost_lgbm:Training until validation scores don't improve for 100 rounds
INFO3:lightautoml.ml_algo.boost_lgbm:[100]	valid's auc: 0.597222
INFO3:lightautoml.ml_algo.boost_lgbm:Early stopping, best iteration is:
[5]	valid's auc: 0.614583
INFO2:lightautoml.ml_algo.base

[02:28:38] Fitting Lvl_0_Pipe_1_Mod_0_Tuned_LightGBM finished. score = 0.6868024553571428


INFO:lightautoml.ml_algo.base:Fitting Lvl_0_Pipe_1_Mod_0_Tuned_LightGBM finished. score = 0.6868024553571428


[02:28:38] Lvl_0_Pipe_1_Mod_0_Tuned_LightGBM fitting and predicting completed


INFO:lightautoml.ml_algo.base:Lvl_0_Pipe_1_Mod_0_Tuned_LightGBM fitting and predicting completed


[02:28:38] Time left 592.30 secs



INFO:lightautoml.automl.base:Time left 592.30 secs



[02:28:38] Layer 1 training completed.



INFO:lightautoml.automl.base:Layer 1 training completed.



[02:28:38] Blending: optimization starts with equal weights. Score = 0.6939174


INFO:lightautoml.automl.blend:Blending: optimization starts with equal weights. Score = 0.6939174


[02:28:38] Blending: iteration 0: score = 0.7134487, weights = [0.6623374  0.33766258]


INFO:lightautoml.automl.blend:Blending: iteration 0: score = 0.7134487, weights = [0.6623374  0.33766258]


[02:28:38] Blending: no improvements for score. Terminated.



INFO:lightautoml.automl.blend:Blending: no improvements for score. Terminated.



[02:28:38] Blending: best score = 0.7134487, best weights = [0.6623374  0.33766258]


INFO:lightautoml.automl.blend:Blending: best score = 0.7134487, best weights = [0.6623374  0.33766258]


[02:28:38] Automl preset training completed in 7.82 seconds



INFO:lightautoml.automl.presets.base:Automl preset training completed in 7.82 seconds



[02:28:38] Model description:
Final prediction for new objects (level 0) = 
	 0.66234 * (5 averaged models Lvl_0_Pipe_0_Mod_0_LinearL2) +
	 0.33766 * (5 averaged models Lvl_0_Pipe_1_Mod_0_Tuned_LightGBM) 



INFO:lightautoml.automl.presets.base:Model description:
Final prediction for new objects (level 0) = 
	 0.66234 * (5 averaged models Lvl_0_Pipe_0_Mod_0_LinearL2) +
	 0.33766 * (5 averaged models Lvl_0_Pipe_1_Mod_0_Tuned_LightGBM) 



[LAMA FS+LGBM/Linear] ROC-AUC: 0.7217  F1: 0.4082


#### Выбор лучшего

In [78]:
results = {}
results['LAMA_DEFAULT'] = (auc1, f1_1)
results['LAMA_FAST'] = (auc2, f1_2)
results['LAMA_ADVANCED'] = (auc3, f1_3)
results['LAMA_FS'] = (auc4, f1_4)
for k, v in results.items():
    print(f'{k:15}| ROC-AUC: {v[0]:.4f} | F1: {v[1]:.4f}')

best_lama_name = max(results, key=lambda k: results[k][0])
print(f'--> Лучший бэйзлайн: {best_lama_name}, ROC-AUC = {results[best_lama_name][0]:.4f}')

LAMA_DEFAULT   | ROC-AUC: 0.7217 | F1: 0.6557
LAMA_FAST      | ROC-AUC: 0.7138 | F1: 0.6552
LAMA_ADVANCED  | ROC-AUC: 0.5000 | F1: 0.0000
LAMA_FS        | ROC-AUC: 0.7217 | F1: 0.4082
--> Лучший бэйзлайн: LAMA_DEFAULT, ROC-AUC = 0.7217


Классическая конфигурация LAMA, несмотря на консерватизм, дала лучший результат среди тестируемых.

### Собственное решение 1

In [62]:
# PIPELINE 1: LightGBM + scale + auto feature selection

# PIPELINE 1A: только scaling + LGBM (без фичер селекции)
pipe_lgbm_base = Pipeline([
    ('scaler', StandardScaler()),
    ('lgbm', LGBMClassifier(n_estimators=100, random_state=42))
])
pipe_lgbm_base.fit(X_train, y_train)
pred_lgbm_base = pipe_lgbm_base.predict_proba(X_test)[:, 1]
auc_lgbm_base = roc_auc_score(y_test, pred_lgbm_base)
print(f'[Pipeline LGBM basic] ROC-AUC: {auc_lgbm_base:.4f}')

# PIPELINE 1B: scaling + feature selection + LGBM
# Селектор тренируется внутри пайплайна по cross-val на train (без data leakage)
pipe_lgbm_fs = Pipeline([
    ('scaler', StandardScaler()),
    ('selector', SelectFromModel(LGBMClassifier(n_estimators=100, random_state=42), threshold="median")),
    ('lgbm', LGBMClassifier(n_estimators=100, random_state=42))
])
pipe_lgbm_fs.fit(X_train, y_train)
pred_lgbm_fs = pipe_lgbm_fs.predict_proba(X_test)[:, 1]
auc_lgbm_fs = roc_auc_score(y_test, pred_lgbm_fs)
print(f'[Pipeline LGBM + feature selection] ROC-AUC: {auc_lgbm_fs:.4f}')

/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but SelectFromModel was fitted with feature names
  warnings.warn(


[Pipeline LGBM basic] ROC-AUC: 0.6980
[Pipeline LGBM + feature selection] ROC-AUC: 0.6742


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but SelectFromModel was fitted with feature names
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LGBMClassifier was fitted with feature names
  warnings.warn(


### Собственное решение 2

In [63]:
# PIPELINE 2: CatBoost + простая обработка пропусков
pipe_catboost = Pipeline([
    ('imputer', SimpleImputer(strategy='mean')),
    ('catboost', CatBoostClassifier(iterations=200,
                                    learning_rate=0.05,
                                    eval_metric='AUC',
                                    random_state=42,
                                    verbose=0))
])

pipe_catboost.fit(X_train, y_train)
pred_cat = pipe_catboost.predict_proba(X_test)[:, 1]
auc_cat = roc_auc_score(y_test, pred_cat)
print(f'[Pipeline CatBoost] ROC-AUC: {auc_cat:.4f}')


[Pipeline CatBoost] ROC-AUC: 0.7048


### Собственное решение 3

In [64]:
# PIPELINE 3: Logistic Regression + StandardScaler + гиперпараметры GridSearch
pipe_logreg = Pipeline([
    ('scaler', StandardScaler()),
    ('logreg', LogisticRegression(max_iter=300))
])

param_grid = {'logreg__C': [0.01, 0.1, 1, 3, 10]}
grid = GridSearchCV(pipe_logreg, param_grid, scoring='roc_auc', cv=5)
grid.fit(X_train, y_train)

print(f"[Pipeline LogReg] Лучшее C: {grid.best_params_}")
best_logreg = grid.best_estimator_
preds_lr = best_logreg.predict_proba(X_test)[:, 1]
auc_lr = roc_auc_score(y_test, preds_lr)
print(f'[Pipeline LogReg tuned] ROC-AUC: {auc_lr:.4f}')

[Pipeline LogReg] Лучшее C: {'logreg__C': 0.1}
[Pipeline LogReg tuned] ROC-AUC: 0.7455


### Вывод
Свое решение оказалось лучше LAMA baseline: максимум ROC-AUC своего решения: 0.7455 (пайплайн 3), лучшего baseline: 0.7217
